sql/src/main/java/io/crate/execution/engine/collect/files/FileReadingIterator.java

```diff
@@ -132,16 +138,19 @@ public void kill(@Nonnull Throwable throwable) {
                                                  Map<String, FileInputFactory> fileInputFactories,
                                                  Boolean shared,
                                                  int numReaders,
-                                                 int readerNumber) {
+                                                 int readerNumber,
+                                                 WriterProjection.InputFormat inputFormat) {
         return new CloseAssertingBatchIterator<>(new FileReadingIterator(fileUris, inputs, collectorExpressions,
-            compression, fileInputFactories, shared, numReaders, readerNumber));
+            compression, fileInputFactories, shared, numReaders, readerNumber, inputFormat));
     }
 
     private void initCollectorState() {
         lineContext = new LineContext();
```

Comment:
I wish we could sublcass lineContext and have a `JsonContext` and a `CsvLineContext`. That way we avoid checking the input for every line and separate concerns.

Reply:
I'd prefer it if we could avoid using inheritance here. Passing a parse method
to the LineContext would do just as well.

Inheritance should be used with caution after evaluating other options first.
It's one of the worst forms of coupling. Having a `LineContext` and subclasses
always raises the question how these classes interact with each other.
Composition on the other hand makes it clear that one component is using the
other. There is no possibility for bi-directional calls. Usually it also ends
up being easier to test as one component stays completely independent of the
other. 

Furthermore, the name `CsvLineContext` imho also wouldn't fit and be
confusing, because the `rawSource` still needs to be in JSON format. (The code
later on in `TransportShardUpsertAction` afaik depends on that being the case)


Ideally we'd go even further and completely decouple any parse logic from this
iterator here. These are separate concerns. E.g. due to the fact that the
BatchIterators are now generic we could model a simple pipeline
`BatchIterator<String> -> BatchIterator<LineContext> -> BatchIterator<Row>`.
(Or go from `String -> Row` at once). This would allow us to separate how we
interpret the lines completely.

Reply:
I agree that the `FileReadingIterator` knows too much here, and shouldn't deal with the differences in parsing csv/json. But I also wouldn't be in favour of going down the inheritance route, if possible here for those same reasons mentioned above. 
I've pulled out the decision for processing the line as csv/json into a separate `LineProcessor` class (name might not be totally appropriate). This takes knowledge of the file type out of the `FileReadingIterator`. However the file type will still be checked for each input line.